guides/client/js-interop.md

```diff
@@ -354,3 +354,47 @@ Hooks.Chart = {
 ```
 
 *Note*: In case a LiveView pushes events and renders content, `handleEvent` callbacks are invoked after the page is updated. Therefore, if the LiveView redirects at the same time it pushes events, callbacks won't be invoked on the old page's elements. Callbacks would be invoked on the redirected page's newly mounted hook elements.
+
+## Locked DOM elements and loading states
+
+Whenever pushings events to the server, such as a button click or form submission, LiveView locks elements in the DOM and applies CSS based loading states as outlined in the [bindings](bindings.md) guide. The lock consists of a `data-phx-ref` attribute, which prevents any inbound patches from
+altering the contents of an element until that ref is acknowledged by the server. This prevents UI race conditions from happenign on the client and provides UI feedback during the round trip.
+
+Custom JavaScript can integrate with the locking events to implement optimistic UI layers, or apply their own custom loading states. The following `CustomEvent`'s are dispatched to an element whenever an event is pushed to the server:
+
+  `phx:lock` – A lock was applied to the element for any phx-binding event.
```

Comment:
do we expect anything outside of a hook to use the locking information? If no, what about a

```
class MyComponent extends HTMLElement {
  connectedCallback(){
    let mounted = () => this.hook.pushEvent(...))
    let lock = (event) => {
      // event.detail = {el, event, phxEvent, ref, loading, lock, unlock}
    }
    this.hook = createHook(this, {mounted, lock})
  }
}

const MyHook = {
  mounted() {
    //
  },
  lock(event) {
    // event.detail = {el, event, phxEvent, ref, loading, lock, unlock}
  }
}
```

I think that feels more familiar to what we're used to when with hooks. Could still be implemented with the events behind the scenes.